File: prototypes/AI_Deltas/LKB/LKB20250314A1B2/src/bias.js
import { config } from "./config.js";
import { clamp } from "./lens.js";

export class BiasField {
  constructor(size, { decay = config.biasDecay, strength = config.biasStrength, radius = config.biasRadius } = {}) {
    this.size = size;
    this.n = size ** 3;
    this.decay = decay;
    this.strength = strength;
    this.radius = radius;
    this.bias = new Float32Array(this.n);
    this.offset = 0;
  }

  tick() {
    for (let i = 0; i < this.n; i++) {
      this.bias[i] *= this.decay;
    }
  }

  injectRadial(center, intensity) {
    const { x, y, z } = center;
    const r = this.radius;
    for (let dz = -r; dz <= r; dz++) {
      for (let dy = -r; dy <= r; dy++) {
        for (let dx = -r; dx <= r; dx++) {
          const d2 = (dx * dx + dy * dy + dz * dz) / (r * r || 1);
          if (d2 > 1) continue;
          const kernel = Math.exp(-d2 * 2.4);
          const idx = this.idx(x + dx, y + dy, z + dz);
          if (idx >= 0) {
            this.bias[idx] = clamp(this.bias[idx] + intensity * this.strength * kernel, -0.3, 0.3);
          }
        }
      }
    }
  }

  idx(x, y, z) {
    if (x < 0 || y < 0 || z < 0 || x >= this.size || y >= this.size || z >= this.size) return -1;
    return x + y * this.size + z * this.size * this.size;
  }

  getBias() {
    return this.bias;
  }

  biasAmplitude() {
    let sum = 0;
    for (let i = 0; i < this.n; i++) sum += Math.abs(this.bias[i]);
    return sum / this.n;
  }
}

File: prototypes/AI_Deltas/LKB/260102193935_88YZJWBB/src/lkb_delta/bias.py
from __future__ import annotations

import math
from typing import Iterable, List

from .config import BiasPulse, SimulationConfig
from .utils import clamp, wrap_index


class BiasField:
    def __init__(self, config: SimulationConfig):
        self.size = config.grid_size
        self.count = config.grid_size ** 3
        self.decay_rate = config.bias_decay
        self.strength = config.bias_strength
        self.default_radius = config.bias_radius
        self.values: List[float] = [0.0 for _ in range(self.count)]

    def _index(self, x: int, y: int, z: int) -> int:
        s = self.size
        return wrap_index(x, s) + wrap_index(y, s) * s + wrap_index(z, s) * s * s

    def decay(self) -> None:
        for i, v in enumerate(self.values):
            self.values[i] = v * self.decay_rate

    def apply_pulse(self, pulse: BiasPulse, strength_scale: float = 1.0) -> None:
        px, py, pz = pulse.position
        radius = max(0.5, pulse.radius or self.default_radius)
        max_dist = math.ceil(radius)
        radius_sq = radius * radius
        for dx in range(-max_dist, max_dist + 1):
            for dy in range(-max_dist, max_dist + 1):
                for dz in range(-max_dist, max_dist + 1):
                    dist_sq = dx * dx + dy * dy + dz * dz
                    if dist_sq > radius_sq:
                        continue
                    kernel = math.exp(-dist_sq / (radius_sq * 0.6))
                    idx = self._index(px + dx, py + dy, pz + dz)
                    delta = pulse.strength * self.strength * kernel * strength_scale
                    self.values[idx] = clamp(self.values[idx] + delta, -0.8, 0.8)

    def apply_pulses(self, pulses: Iterable[BiasPulse], current_step: int) -> None:
        for pulse in pulses:
            if pulse.step == current_step:
                self.apply_pulse(pulse)

    def amplitude(self) -> float:
        return max(abs(v) for v in self.values) if self.values else 0.0

    def flatten(self) -> List[float]:
        return list(self.values)
